File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int Max=Integer.MIN_VALUE;
    int val(TreeNode root)
    {
        if(root==null)
        {
            return 0;
        }

        int leftVal= Math.max(val(root.left), 0);
        int rightVal= Math.max(val(root.right),0);

        int currVal=root.val+leftVal+rightVal;
        Max=Math.max(currVal,Max);


        return root.val+Math.max(leftVal,rightVal);
    }
    public int maxPathSum(TreeNode root) {
        val(root);

        return Max;
    }
}

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxSum=INT_MIN;
    int val(TreeNode* root)
    {
        if(root==NULL)
        return 0;

        int leftVal=max(val(root->left),0);
        int rightVal=max(val(root->right),0);

        int currentPathSum=root->val+leftVal+rightVal;

        maxSum=max(maxSum, currentPathSum);

        return root->val+max(leftVal,rightVal);
         
    }
    int maxPathSum(TreeNode* root) {
        
        if(root==NULL)return 0;
        val(root);
        return maxSum;
    }
};